digits/framework_helpers/caffe_helpers.py

```diff
@@ -0,0 +1,269 @@
+            # Check to see if num_output is unset
+            if not layer.inner_product_param.HasField('num_output'):
+                layer.inner_product_param.num_output = num_categories
+
+    return network
+
+def cleanedUpGenericNetwork(original_network):
+    """
+    Perform a few cleanup routines on a generic network
+    Returns a new NetParameter
+    """
+    network = caffe_pb2.NetParameter()
+    network.CopyFrom(original_network)
+
+    for i, layer in enumerate(network.layer):
+        if 'Data' in layer.type:
+            assert layer.type in ['Data'], \
+                'Unsupported data layer type %s' % layer.type
+
+        elif layer.type == 'Input':
+            # DIGITS handles the deploy file for you
+            del network.layer[i]
+
+        elif layer.type == 'InnerProduct':
+            # Check to see if num_output is unset
+            assert layer.inner_product_param.HasField('num_output'), \
+                "Don't leave inner_product_param.num_output unset for generic networks (layer %s)" % layer.name
+
+    return network
+
+def _layerIncludedInState(layer, state):
+    """
+    Returns True if this layer will be included in the given state
+    Logic copied from Caffe's Net::FilterNet()
+    """
+    # If no include rules are specified, the layer is included by default and
+    # only excluded if it meets one of the exclude rules.
+    layer_included = len(layer.include) == 0
+
+    for exclude_rule in layer.exclude:
+        if _stateMeetsRule(state, exclude_rule):
+            layer_included = False
+            break
+
+    for include_rule in layer.include:
+        if _stateMeetsRule(state, include_rule):
+            layer_included = True
+            break
+
+    return layer_included
+
+def _stateMeetsRule(state, rule):
+    """
+    Returns True if the given state meets the given rule
+    Logic copied from Caffe's Net::StateMeetsRule()
+    """
+    if rule.HasField('phase'):
+        if rule.phase != state.phase:
+            return False
+
+    if rule.HasField('min_level'):
+        if state.level < rule.min_level:
+            return False
+
+    if rule.HasField('max_level'):
+        if state.level > rule.max_level:
+            return False
+
+    # The state must contain ALL of the rule's stages
+    for stage in rule.stage:
+        if stage not in state.stage:
+            return False
+
+    # The state must contain NONE of the rule's not_stages
+    for stage in rule.not_stage:
+        if stage in state.stage:
+            return False
+
+    return True
+
+def _setLayerRule(layer, rule=None):
+    """
+    Set a new include rule for this layer
+    If rule is None, the layer will always be included
+    """
+    layer.ClearField('include')
+    layer.ClearField('exclude')
+    if rule is not None:
+        layer.include.add().CopyFrom(rule)
+
+def save_deploy_file_classification(network,path,num_categories, crop_size=None, feature_dims=None,logger=None):
```

Comment:
@gheinrich Thats a good point! Since pretrained models can be from outside of digits I didn't include it in the upload process, plus they are also used for fine tuning. Although perhaps we could have check if a deploy file exists, and if so ignore the generate deploy step?